toolchain/check/control_flow.cpp

```diff
@@ -130,4 +135,56 @@ auto IsCurrentPositionReachable(Context& context) -> bool {
          SemIR::TerminatorKind::Terminator;
 }
 
+auto MaybeAddCleanupForInst(Context& context, SemIR::TypeId type_id,
+                            SemIR::InstId inst_id) -> void {
+  if (!context.scope_stack().PeekIsLexicalScope()) {
+    // Cleanup can only occur in lexical scopes.
+    return;
+  }
+
+  // TODO: Add destruction of members of ArrayType, StructType, and
+  // TupleType. Includes refactoring MaybeAddCleanupForInst to remain
+  // non-recursive.
+
+  auto type_inst = context.types().GetAsInst(type_id);
+  CARBON_KIND_SWITCH(type_inst) {
+    case SemIR::ClassType::Kind: {
+      // TODO: Figure out what destruction of classes with destroyable members
+      // should look like (maybe an implicit `fn destroy` added by
+      // `handle_class.cpp`?).
+      auto destroy_id =
+          PerformMemberAccess(context, SemIR::LocId::None, inst_id,
+                              SemIR::NameId::Destroy, /*required=*/false);
+      context.scope_stack().destroy_id_stack().AppendToTop(destroy_id);
+      break;
+    }
+
+    default:
+      // Not interesting storage for destruction.
+      return;
+  }
+}
+
+// Common support for cleanup blocks.
+static auto AddCleanupBlock(Context& context) -> void {
+  auto destroy_ids = context.scope_stack().destroy_id_stack().PeekArray();
+
+  // If there's nothing to destroy, add the final instruction to the current
+  // block.
+  if (destroy_ids.empty()) {
+    return;
+  }
+
+  for (auto destroy_id : destroy_ids) {
```

Comment:
Should we do this in reverse order, so that the most recently created object is destroyed first?

Reply:
Oops, fixed. In my head this was a stack and already doing that -- obviously not.